### PR TITLE
ARROW-17980: [C++] As-of-Join Substrait extension

### PR DESCRIPTION
**Pull request by rtpsw**
Replacing https://github.com/apache/arrow/pull/14385

### CODE REVIEW

File: cpp/src/arrow/compute/exec/asof_join_benchmark.cc
Comment (rtpsw) [inline cpp/src/arrow/compute/exec/asof_join_benchmark.cc:78]: This line and similar ones seem like an unintended change.

File: cpp/src/arrow/engine/substrait/test_plan_builder.cc
Comment (rtpsw) [inline cpp/src/arrow/engine/substrait/test_plan_builder.cc:94]: This change seems unintended.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (westonpace) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:665]: Hmm...this isn't great to be relying on a Substrait fork.  That being said, we aren't going to be compatible with anything until https://github.com/substrait-io/substrait/pull/342 merges.  So I suppose this is just a natural consequence of working with emerging specifications.  It puts us in a better state for the short term.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (westonpace) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1721]:
```suggestion
```
This was just a debugging message

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (westonpace) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1773]:
```suggestion
```

File: cpp/proto/substrait/extension_rels.proto
Comment (westonpace) [inline cpp/proto/substrait/extension_rels.proto:19]: I'm not a big fan of needing all of the:
```
namespace substrait = ::substrait;
```

but I suppose we can live with it if you are determined to have this package name.  Did `package arrow.substrait_ext` not work?

Could there be another alternative?  Would it be tolerable to have `package substrait.arrow` instead?  It's somewhat unfortunate that protobuf doesn't have options for controlling the C++ namespace like they do for Java & C#.

File: cpp/src/arrow/engine/substrait/plan_internal.cc
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/plan_internal.cc:140]:
```suggestion
// TODO(ARROW-18145) Populate these from cmake files
```

File: cpp/src/arrow/engine/substrait/test_plan_builder.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/test_plan_builder.cc:94]: It's part of adapting to the new options format.  Function arguments can no longer be unspecified.

File: cpp/src/arrow/engine/substrait/util.cc
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/util.cc:128]: We should revert this change.

File: cpp/src/arrow/compute/exec/asof_join_benchmark.cc
Comment (westonpace) [reply cpp/src/arrow/compute/exec/asof_join_benchmark.cc:78]: I think these are leftover from your brief merge on top of the PR that adds the proper single-threaded execution mode.  In that PR `ExecPlan::Make` no longer takes an exec context since the exec context has a thread pool and we don't want to actually provide or start using the thread pool until we start producing.  It's somewhat harmless here since `ctx` is identical to the default exec context.  However, it is not harmless in `substrait/util.cc` because the context there is actually using the CPU thread pool.

**Review by westonpace — COMMENTED** [review]
I'm a little torn on this PR.  It seems that it includes https://github.com/apache/arrow/pull/14415 which hasn't merged yet, although that is primarily because it blocked waiting for https://github.com/substrait-io/substrait/pull/342 .  At this point I think @bkietz remaining comments were fairly minor and I had addressed them.  Also, since things are a bit squashed, it is unclear to me how much of #14415 this includes.  Do you know if you were able to squash in all the commits from #14415 ?

Also, this effectively means we are temporarily based on a Substrait PR while we wait for Substrait to catch up.

I'm not a big fan of the fact that we need the using statements for the substrait namespace but if we really want that package name then it seems protobuf gives us little choice.

That being said, this PR works, I don't see anything functionally wrong with it, it is moving in the right direction, and it will make adopting the official Substrait release a smaller task.  We should probably proceed by merging in https://github.com/apache/arrow/pull/14415 properly, instead of forcing it through integrated as part of this PR though.

File: cpp/proto/substrait/extension_rels.proto
Comment (rtpsw) [reply cpp/proto/substrait/extension_rels.proto:19]: In an effort to get some version to work, I ended up not checking whether `arrow.substrait_ext` worked, but I see no reason why it shouldn't. I ended up finding a small/local change that solves the problem MSVC had with looking up a symbol in the protobuf `substrait` namespace after not finding one in Arrow's.

File: cpp/src/arrow/engine/substrait/options.h
Comment (anjakefala) [inline cpp/src/arrow/engine/substrait/options.h:26]: This would mean that libarrow_python would require protobuf headers to build.

Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/options.h:26]: I'm not familiar with libarrow_python build, but it sounds like you're saying adding this would be bad. Should the APIs involving `Any` be made internal?

**Review by anjakefala — COMMENTED** [review]
I left one comment about whether we want to require protobuf headers in order to build libarrow_python with substrait.

Additionally, find_package(Protobuf) seems to be missing from the CMakeLists

I would add the following here https://github.com/rtpsw/arrow/blob/fce4dff7df8577731cccb220457ade5e24a33e60/python/CMakeLists.txt#L232


```
if(ARROW_PROTOBUF_USE_SHARED)
  find_package(Protobuf)
  include_directories(${PROTOBUF_INCLUDE_DIRS})
  get_property(dirs DIRECTORY PROPERTY INCLUDE_DIRECTORIES)
  foreach(dir ${dirs})
    message("dir='${dir}'")
  endforeach()
  get_cmake_property(_variableNames VARIABLES)
  foreach (_variableName ${_variableNames})
    message("${_variableName}=${${_variableName}}")
  endforeach()
endif()
```

But, there might be a better place for it.

**Comment by westonpace** [comment]
> I think before merging we should address if this is behaviour we want to keep.

@anjakefala It is not.  I would like to hide protobuf from python (I suspect an options_internal.h is doable somehow).

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (westonpace) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:666]: This will need a rebase now that #14415 has merged.  I'm not sure if this block will merge cleanly but it can be removed.

File: cpp/proto/substrait/extension_rels.proto
Comment (westonpace) [reply cpp/proto/substrait/extension_rels.proto:19]: @rtpsw I believe, in our experiment branch, we (and ibis-substrait) ended up going with `package arrow.substrait_ext;`  Do you want to update this here (and if so we can remove the extra `namespace` statements.  If not, we will need to update ibis-substrait I think.

File: cpp/src/arrow/compute/exec/options.h
Comment (westonpace) [inline cpp/src/arrow/compute/exec/options.h:405]:
```suggestion
    /// type. Inexact match is used on the "on" key. i.e., a row is considered a match iff
```

File: cpp/src/arrow/compute/exec/options.h
Comment (westonpace) [inline cpp/src/arrow/compute/exec/options.h:406]: Is the right side really `left_on` or is it `left_on + tolerance`?

File: cpp/src/arrow/compute/exec/options.h
Comment (westonpace) [inline cpp/src/arrow/compute/exec/options.h:411]:
```suggestion
    /// Each input table must have each field of the "by" key.  Exact equality is used for
```

File: cpp/src/arrow/compute/exec/options.h
Comment (westonpace) [inline cpp/src/arrow/compute/exec/options.h:414]: Do we support bool here now?  Or is that still only for payloads?

File: cpp/src/arrow/compute/exec/options.h
Comment (westonpace) [inline cpp/src/arrow/compute/exec/options.h:423]:
```suggestion
  /// \see `Keys` for details.
```

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/expression_internal.cc:949]: The changes in this file are probably out of date.  If you hit merge issues I'm pretty sure you can just discard these changes in this file.

File: cpp/src/arrow/engine/substrait/options.cc
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/options.cc:107]: There is a proposal on https://github.com/ibis-project/ibis-substrait/issues/423 to fix up the output ordering to something ibis-substrait expects.  I think we can handle it as part of this PR or as a separate follow-up PR.  Let me know which you'd prefer.

**Review by westonpace — CHANGES_REQUESTED** [review]
Now that #14415 has merged we can move forwards on this.  However, the code here is out of date (it was based on an older revision of #14415) and I worry it might not merge so cleanly.  Let me know if you have trouble and I can take a stab at it.

**Comment by rtpsw** [comment]
> Now that #14415 has merged we can move forwards on this. However, the code here is out of date (it was based on an older revision of #14415) and I worry it might not merge so cleanly. Let me know if you have trouble and I can take a stab at it.

I will try to create a replacement PR.

**Comment by rtpsw** [comment]
I ended up merging. @westonpace, does it look like what you expected?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (rtpsw) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Done.

File: cpp/src/arrow/compute/exec/options.h
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/options.h:458]: Yes, it is `left_on`, because the right tables are never scanned beyond the `on` value seen in the left table; the meaning of "as-of" is up to this value.

File: cpp/src/arrow/compute/exec/options.h
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/options.h:466]: Currently, `AsofJoinNode` support `bool` for data (payload) fields only. It is not too hard to add support for `bool` on-key fields, but let's do it separately.

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/expression_internal.cc]: Merge seems OK.

File: cpp/src/arrow/engine/substrait/options.cc
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/options.cc:107]: I'd prefer a separate follow-up PR.

File: cpp/proto/substrait/extension_rels.proto
Comment (rtpsw) [reply cpp/proto/substrait/extension_rels.proto]: Fixed.

**Comment by rtpsw** [comment]
How to deal with the following [CI job's](https://github.com/apache/arrow/actions/runs/3557712768/jobs/5975904381) error?
```
/arrow/cpp/src/arrow/engine/substrait/options.h:37: error: documented symbol 'enum ARROW_ENGINE_EXPORT arrow::engine::arrow::engine::ConversionStrictness' was not declared or defined. (warning treated as error, aborting now)
```

**Comment by rtpsw** [comment]
> I'm not a big fan of the fact that we need the using statements for the substrait namespace but if we really want that package name then it seems protobuf gives us little choice.

@westonpace, note that this PR is currently using a `substrait_ext` namespace

**Comment by westonpace** [comment]
@rtpsw Looks like the merge didn't go through cleanly.  Let me know if you want me to clean this up.  Have you been able to address the concern here: https://github.com/apache/arrow/pull/14485#pullrequestreview-1167652534 ?

**Comment by rtpsw** [comment]
> @rtpsw Looks like the merge didn't go through cleanly. Let me know if you want me to clean this up.

@westonpace, if you can easily clean this up, that would be helpful.

> Have you been able to address the concern here: [#14485 (review)](https://github.com/apache/arrow/pull/14485#pullrequestreview-1167652534) ?

I have removed the dependence on protobuf in the header, but I haven't added the proposed cmake snippet. It sounded like the place for this snippet is under discussion. I'm also not sure whether it is required in this PR or can be deferred.

**Review by westonpace — APPROVED** [review]
@rtpsw I had to rebase this again.  I've looked through it and I think this is in good shape now.  If you're ok with it I will merge once the tests pass.

**Comment by rtpsw** [comment]
Most CI failures seem unrelated, in particular this Substrait-test failure:
```
D:/a/arrow/arrow/cpp/src/arrow/status.cc:137: When trying to delete temporary directory: IOError: Cannot delete file 'C:/Users/RUNNER~1/AppData/Local/Temp/substrait-tempdir-q1gzn3f7/serde_test.arrow': . Detail: [Windows error 32] The process cannot access the file because it is being used by another process.
...
[  FAILED  ] 1 test, listed below:
[  FAILED  ] Substrait.ReadRelWithGlobFiles
```
but I'm not sure about [this one](https://github.com/apache/arrow/actions/runs/3662217235/jobs/6191152672):
```
[ RUN      ] TestExpressionRegistry.VerifySupportedFunctions
/Users/runner/work/arrow/arrow/cpp/src/gandiva/expression_registry_test.cc:48: Failure
Expected: (element) != (functions.end()), actual: 8-byte object <40-DB 8C-8D 9B-7F 00-00> vs 8-byte object <40-DB 8C-8D 9B-7F 00-00>
function signature date64[ms] date_diff(date64[ms], int64) missing in supported functions.

/Users/runner/work/arrow/arrow/cpp/src/gandiva/expression_registry_test.cc:48: Failure
Expected: (element) != (functions.end()), actual: 8-byte object <40-DB 8C-8D 9B-7F 00-00> vs 8-byte object <40-DB 8C-8D 9B-7F 00-00>
function signature timestamp[ms] date_diff(timestamp[ms], int64) missing in supported functions.

[  FAILED  ] TestExpressionRegistry.VerifySupportedFunctions (30 ms)
```

**Comment by rtpsw** [comment]
> @rtpsw I had to rebase this again. I've looked through it and I think this is in good shape now. If you're ok with it I will merge once the tests pass.

I compared the codes, mostly manually, and couldn't find anything missing.

**Comment by westonpace** [comment]
The tests were failing after a rebase due to a recent change to CheckRoundTrip (doesn't need a schema).  I pushed a small fix.

Unfortunately, the test is now failing for me because the process thread marks the plan complete and then tries to join itself.  I copied over the change we made in https://github.com/apache/arrow/commit/b9cda43bedb70a1e64a60ad712c58f471a7bc13d to address this but that change relies on an executor always being present which won't be true until ARROW-15732 merges.

**Comment by rtpsw** [comment]
@westonpace, I see you approved - is there anything holding up this PR?

**Comment by westonpace** [comment]
@rtpsw I believe it's the same failure you noticed before, which is that there is a segfault because schedule task is synchronous.  In the other PR  you fixed it by requiring an executor but that introduces ordering problems if I remember correctly.  Although it may not matter for a test and will allow us to move forward on this.  I will try this real quick.

**Comment by rtpsw** [comment]
> @rtpsw I believe it's the same failure you noticed before, which is that there is a segfault because schedule task is synchronous. In the other PR you fixed it by requiring an executor but that introduces ordering problems if I remember correctly. Although it may not matter for a test and will allow us to move forward on this. I will try this real quick.

If indeed it introduces these ordering problems, there are alternatives fixes for the segfault, e.g., use a standalone thread for that task or use a new `ExecContext::async_executor()` method that is guaranteed to return an asynchronous executor.

**Comment by westonpace** [comment]
> If indeed it introduces these ordering problems, there are alternatives fixes for the segfault, e.g., use a standalone thread for that task or use a new ExecContext::async_executor() method that is guaranteed to return an asynchronous executor.

Agreed.  I think we can address this later though.  I've rebased (had to address the query context change) and would like to merge this when it is green.

**Comment by westonpace** [comment]
> Agreed. I think we can address this later though. I've rebased (had to address the query context change) and would like to merge this when it is green.

The problem was still happening.  I've fixed it properly in #15104 using the same fix that was in the demo branch (ensuring that we always have an executor and the async scheduler is never synchronous).

I've rebased this (once more) and the problem *should* no longer occur.

**Comment by westonpace** [comment]
CI failure seems surprising but unrelated.  I filed #15137 to follow up.

**Comment by westonpace** [comment]
Failure appears unrelated.  I will merge.

**Comment by westonpace** [comment]
@rtpsw thanks for your persistence on this one.  Sorry for the rebase troubles at the end.

**Comment by rtpsw** [comment]
@westonpace, thanks for wrapping this up.

**Pull request merged**
